File: android/app/src/main/java/com/example/bonermohis/data/CalculationsHelper.kt
package com.example.bonermohis.data

import kotlin.math.max
import kotlin.math.min

data class Slab(val limit: Double, val rate: Double)

data class SlabBreakdownLine(
    val name: String,
    val units: Double,
    val rate: Double,
    val cost: Double
)

data class SlabDetails(
    val index: Int,
    val rate: Double,
    val slabMin: Double,
    val slabMax: Double,
    val percentage: Double,
    val label: String
)

data class DistributorConfig(
    val name: String,
    val currency: String,
    val slabs: List<Slab>
)

object CalculationsHelper {
    
    fun getSlabBreakdown(kwh: Double, provider: String = "default"): List<SlabBreakdownLine> {
        val config = DISTRIBUTORS[provider] ?: DISTRIBUTORS["default"]!!
        var remaining = kwh
        val lines = mutableListOf<SlabBreakdownLine>()

        val isLifeline = kwh <= 50.0
        val startIdx = if (isLifeline) 0 else 1
        var prevLimit = 0.0

        val slabNames = listOf(
            "Lifeline", "First Step", "Second Step", "Third Step", 
            "Fourth Step", "Fifth Step", "Sixth Step"
        )

        for (i in startIdx until config.slabs.size) {
            val slab = config.slabs[i]
            val rangeWidth = slab.limit - prevLimit
            val units = min(remaining, rangeWidth)

            if (units > 0.0) {
                val cost = units * slab.rate
                val name = slabNames.getOrNull(i) ?: "Slab ${i + 1}"
                lines.add(
                    SlabBreakdownLine(
                        name = name,
                        units = Math.round(units * 100.0) / 100.0,
                        rate = slab.rate,
                        cost = Math.round(cost * 100.0) / 100.0
                    )
                )
            }

            remaining -= units
            prevLimit = slab.limit
            if (remaining <= 0.0) break
        }

        return lines
    }

    val DISTRIBUTORS = mapOf(
        "default" to DistributorConfig(
            name = "Standard Progressive Utility",
            currency = "৳",
            slabs = listOf(
                Slab(50.0, 4.63),
                Slab(75.0, 5.26),
                Slab(200.0, 8.50),
                Slab(300.0, 9.10),
                Slab(400.0, 9.62),
                Slab(600.0, 15.01),
                Slab(Double.POSITIVE_INFINITY, 17.35)
            )
        ),
        "dpdc" to DistributorConfig(
            name = "DPDC (Dhaka Power)",
            currency = "৳",
            slabs = listOf(
                Slab(50.0, 4.63),
                Slab(75.0, 5.26),
                Slab(200.0, 8.50),
                Slab(300.0, 9.10),
                Slab(400.0, 9.62),
                Slab(600.0, 15.01),
                Slab(Double.POSITIVE_INFINITY, 17.35)
            )
        ),
        "desco" to DistributorConfig(
            name = "DESCO (Dhaka Electric)",
            currency = "৳",
            slabs = listOf(
                Slab(50.0, 4.63),
                Slab(75.0, 5.26),
                Slab(200.0, 8.50),
                Slab(300.0, 9.10),
                Slab(400.0, 9.62),
                Slab(600.0, 15.01),
                Slab(Double.POSITIVE_INFINITY, 17.35)
            )
        )
    )

    fun calculateCost(kwh: Double, provider: String = "default"): Double {
        val config = DISTRIBUTORS[provider] ?: DISTRIBUTORS["default"]!!
        var remaining = kwh
        var totalCost = 0.0
        
        val isLifeline = kwh <= 50.0
        val startIdx = if (isLifeline) 0 else 1
        var prevLimit = 0.0

        for (i in startIdx until config.slabs.size) {
            val slab = config.slabs[i]
            val rangeWidth = slab.limit - prevLimit
            val consumedInSlab = min(remaining, rangeWidth)
            totalCost += consumedInSlab * slab.rate
            remaining -= consumedInSlab
            prevLimit = slab.limit
            if (remaining <= 0.0) break
        }

        return Math.round(totalCost * 100.0) / 100.0
    }

    fun getSlabDetails(kwh: Double, provider: String = "default"): SlabDetails {
        val config = DISTRIBUTORS[provider] ?: DISTRIBUTORS["default"]!!
        
        // Find which slab index the current kwh falls into
        var index = 0
        for (i in config.slabs.indices) {
            if (kwh <= config.slabs[i].limit || config.slabs[i].limit == Double.POSITIVE_INFINITY) {
                index = i
                break
            }
        }

        val slab = config.slabs[index]

        data class SlabRange(val name: String, val min: Double, val max: Double)
        val slabRanges = listOf(
            SlabRange("Lifeline", 0.0, 50.0),
            SlabRange("First Step", if (kwh > 50.0) 0.0 else 51.0, 75.0),
            SlabRange("Second Step", 76.0, 200.0),
            SlabRange("Third Step", 201.0, 300.0),
            SlabRange("Fourth Step", 301.0, 400.0),
            SlabRange("Fifth Step", 401.0, 600.0),
            SlabRange("Sixth Step", 601.0, Double.POSITIVE_INFINITY)
        )

        val rangeConfig = slabRanges.getOrElse(index) {
            SlabRange("Slab ${index + 1}", 0.0, slab.limit)
        }

        val slabMin = rangeConfig.min
        val slabMax = rangeConfig.max
        val rangeWidth = slabMax - slabMin

        val percentage = if (rangeWidth == Double.POSITIVE_INFINITY) {
            100.0
        } else {
            min(100.0, max(0.0, ((kwh - slabMin) / rangeWidth) * 100.0))
        }

        val label = if (slabMax == Double.POSITIVE_INFINITY) {
            "${rangeConfig.name} (> ${(slabMin - 1).toInt()} kWh)"
        } else {
            "${rangeConfig.name} (${slabMin.toInt()}-${slabMax.toInt()} kWh)"
        }

        return SlabDetails(
            index = index,
            rate = slab.rate,
            slabMin = slabMin,
            slabMax = slabMax,
            percentage = Math.round(percentage * 10.0) / 10.0,
            label = label
        )
    }

    fun calculateDaysRemaining(balance: Double, yesterdayUsage: Double): Double {
        if (yesterdayUsage <= 0.0) return Double.POSITIVE_INFINITY
        return Math.round((balance / yesterdayUsage) * 10.0) / 10.0
    }
}

File: lib/data/calculations_helper.dart
import 'dart:math';
import 'models/slab_models.dart';

/// Port of CalculationsHelper.kt — BPDB Category-A Residential 7-slab tariff.
/// See docs/TARIFF.md for full specification and lifeline bypass rule.
class CalculationsHelper {
  CalculationsHelper._();

  static const Map<String, DistributorConfig> distributors = {
    'default': DistributorConfig(
      name: 'Standard Progressive Utility',
      currency: '৳',
      slabs: [
        Slab(50.0, 4.63),
        Slab(75.0, 5.26),
        Slab(200.0, 8.50),
        Slab(300.0, 9.10),
        Slab(400.0, 9.62),
        Slab(600.0, 15.01),
        Slab(double.infinity, 17.35),
      ],
    ),
    'dpdc': DistributorConfig(
      name: 'DPDC (Dhaka Power)',
      currency: '৳',
      slabs: [
        Slab(50.0, 4.63),
        Slab(75.0, 5.26),
        Slab(200.0, 8.50),
        Slab(300.0, 9.10),
        Slab(400.0, 9.62),
        Slab(600.0, 15.01),
        Slab(double.infinity, 17.35),
      ],
    ),
    'desco': DistributorConfig(
      name: 'DESCO (Dhaka Electric)',
      currency: '৳',
      slabs: [
        Slab(50.0, 4.63),
        Slab(75.0, 5.26),
        Slab(200.0, 8.50),
        Slab(300.0, 9.10),
        Slab(400.0, 9.62),
        Slab(600.0, 15.01),
        Slab(double.infinity, 17.35),
      ],
    ),
  };

  static const _slabNames = [
    'Lifeline',
    'First Step',
    'Second Step',
    'Third Step',
    'Fourth Step',
    'Fifth Step',
    'Sixth Step',
  ];

  static double calculateCost(double kwh, {String provider = 'default'}) {
    final config = distributors[provider] ?? distributors['default']!;
    double remaining = kwh;
    double totalCost = 0.0;

    final isLifeline = kwh <= 50.0;
    final startIdx = isLifeline ? 0 : 1;
    double prevLimit = 0.0;

    for (int i = startIdx; i < config.slabs.length; i++) {
      final slab = config.slabs[i];
      final rangeWidth = slab.limit - prevLimit;
      final consumedInSlab = min(remaining, rangeWidth);
      totalCost += consumedInSlab * slab.rate;
      remaining -= consumedInSlab;
      prevLimit = slab.limit;
      if (remaining <= 0.0) break;
    }

    return (totalCost * 100).roundToDouble() / 100.0;
  }

  static SlabDetails getSlabDetails(double kwh,
      {String provider = 'default'}) {
    final config = distributors[provider] ?? distributors['default']!;

    int index = 0;
    for (int i = 0; i < config.slabs.length; i++) {
      if (kwh <= config.slabs[i].limit ||
          config.slabs[i].limit == double.infinity) {
        index = i;
        break;
      }
    }

    final slab = config.slabs[index];

    final slabRanges = [
      const _SlabRange('Lifeline', 0.0, 50.0),
      _SlabRange('First Step', kwh > 50.0 ? 0.0 : 51.0, 75.0),
      const _SlabRange('Second Step', 76.0, 200.0),
      const _SlabRange('Third Step', 201.0, 300.0),
      const _SlabRange('Fourth Step', 301.0, 400.0),
      const _SlabRange('Fifth Step', 401.0, 600.0),
      const _SlabRange('Sixth Step', 601.0, double.infinity),
    ];

    final rangeConfig = index < slabRanges.length
        ? slabRanges[index]
        : _SlabRange('Slab ${index + 1}', 0.0, slab.limit);

    final slabMin = rangeConfig.min;
    final slabMax = rangeConfig.max;
    final rangeWidth = slabMax - slabMin;

    final percentage = rangeWidth == double.infinity
        ? 100.0
        : min(100.0, max(0.0, ((kwh - slabMin) / rangeWidth) * 100.0));

    final label = slabMax == double.infinity
        ? '${rangeConfig.name} (>${(slabMin - 1).toInt()} kWh)'
        : '${rangeConfig.name} (${slabMin.toInt()}–${slabMax.toInt()} kWh)';

    return SlabDetails(
      index: index,
      rate: slab.rate,
      slabMin: slabMin,
      slabMax: slabMax,
      percentage: (percentage * 10).roundToDouble() / 10.0,
      label: label,
    );
  }

  static List<SlabBreakdownLine> getSlabBreakdown(double kwh,
      {String provider = 'default'}) {
    final config = distributors[provider] ?? distributors['default']!;
    double remaining = kwh;
    final lines = <SlabBreakdownLine>[];

    final isLifeline = kwh <= 50.0;
    final startIdx = isLifeline ? 0 : 1;
    double prevLimit = 0.0;

    for (int i = startIdx; i < config.slabs.length; i++) {
      final slab = config.slabs[i];
      final rangeWidth = slab.limit - prevLimit;
      final units = min(remaining, rangeWidth);

      if (units > 0.0) {
        final cost = units * slab.rate;
        final name = i < _slabNames.length ? _slabNames[i] : 'Slab ${i + 1}';
        lines.add(SlabBreakdownLine(
          name: name,
          units: (units * 100).roundToDouble() / 100.0,
          rate: slab.rate,
          cost: (cost * 100).roundToDouble() / 100.0,
        ));
      }

      remaining -= units;
      prevLimit = slab.limit;
      if (remaining <= 0.0) break;
    }

    return lines;
  }

  static double calculateDaysRemaining(
      double balance, double yesterdayUsage) {
    if (yesterdayUsage <= 0.0) return double.infinity;
    return (balance / yesterdayUsage * 10).roundToDouble() / 10.0;
  }
}

class _SlabRange {
  final String name;
  final double min;
  final double max;
  const _SlabRange(this.name, this.min, this.max);
}
